### ISSUE TITLE
TFDS fails when running distributed

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/datasets#1302]
**Short description**
When running TF with multiple workers (e.g. using `MultiWorkerMirroredStrategy`) then all workers seem to try and download & extract the dataset. This will then fail in the process due to accessing the same files

**Environment information**
* Operating System: RHEL 7.4
* Python version: 3.7.4
* `tensorflow-datasets`/`tfds-nightly` version: tensorflow-datasets 1.3.2
* `tensorflow`/`tensorflow-gpu`/`tf-nightly`/`tf-nightly-gpu` version: tensorflow 2.0.0

**Reproduction instructions**

```
import tensorflow as tf
import tensorflow_datasets as tfds
strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy()
datasets, info = tfds.load(name='stanford_dogs',
                               with_info=True,
                               as_supervised=True,
                               shuffle_files=False)
```

**Link to logs**
```
Traceback (most recent call last):
  File "git/tensorflow_tests/train_distributed.py", line 216, in <module>
    run()  #pylint: disable=no-value-for-parameter
  File " <pythondir>/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File " <pythondir>/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File " <pythondir>/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File " <pythondir>/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "git/tensorflow_tests/train_distributed.py", line 122, in run
    shuffle_files=False)
  File " <pythondir>/site-packages/tensorflow_datasets/core/api_utils.py", line 52, in disallow_positional_args_dec
    return fn(*args, **kwargs)
  File " <pythondir>/site-packages/tensorflow_datasets/core/registered.py", line 302, in load
    dbuilder.download_and_prepare(**download_and_prepare_kwargs)
  File " <pythondir>/site-packages/tensorflow_datasets/core/api_utils.py", line 52, in disallow_positional_args_dec
    return fn(*args, **kwargs)
  File " <pythondir>/site-packages/tensorflow_datasets/core/dataset_builder.py", line 322, in download_and_prepare
    download_config=download_config)
  File " <pythondir>/site-packages/tensorflow_datasets/core/dataset_builder.py", line 969, in _download_and_prepare
    max_examples_per_split=download_config.max_examples_per_split,
  File " <pythondir>/site-packages/tensorflow_datasets/core/dataset_builder.py", line 825, in _download_and_prepare
    for split_generator in self._split_generators(dl_manager):
  File " <pythondir>/site-packages/tensorflow_datasets/image/stanford_dogs.py", line 133, in _split_generators
    xml_file_list[fname] = ET.parse(f)
  File " <pythondir>/xml/etree/ElementTree.py", line 1197, in parse
    tree.parse(source, parser)
  File " <pythondir>/xml/etree/ElementTree.py", line 598, in parse
    self._root = parser._parse_whole(source)
  File " <pythondir>/site-packages/tensorflow_core/python/lib/io/file_io.py", line 122, in read
    self._preread_check()
  File " <pythondir>/site-packages/tensorflow_core/python/lib/io/file_io.py", line 84, in _preread_check
    compat.as_bytes(self.__name), 1024 * 512)
tensorflow.python.framework.errors_impl.NotFoundError: /home/s3248973/tensorflow_datasets/downloads/extracted/TAR.vision.stanfor.edu_aditya8_ImageNe_annotatZi7Tb3B75rGVwv2jktjhU-g68oztyOITSmSfGY3vMP0.tar/Anno
```

**Expected behavior**
Should work out of the box and e.g. only use 1 worker to download

**Additional context**
started with e.g. `srun --ntasks=2` and appropriately set `TF_CONFIG` env var

**Comment by Flamefire** [comment]
Putting the `load` into `with strategy.scope()` leads to: 

```
Downloading and preparing dataset stanford_dogs (778.12 MiB) to /home/s3248973/tensorflow_datasets/stanford_dogs/0.2.0...

Shuffling and writing examples to /home/s3248973/tensorflow_datasets/stanford_dogs/0.2.0.incomplete63C694/stanford_dogs-train.tfrecord
Shuffling and writing examples to /home/s3248973/tensorflow_datasets/stanford_dogs/0.2.0.incomplete63C694/stanford_dogs-test.tfrecord [Yes, this is duplicated]
[Similar as above]
  File "/home/s3248973/.local/lib/python3.7/site-packages/tensorflow_datasets/core/dataset_builder.py", line 342, in download_and_prepare
    self.info.write_to_directory(self._data_dir)
  File "<pythondir>/site-packages/tensorflow_datasets/core/file_format_adapter.py", line 200, in incomplete_dir
    tf.io.gfile.rename(tmp_dir, dirname)
  File "<pythondir>/site-packages/tensorflow_core/python/lib/io/file_io.py", line 519, in rename_v2
    compat.as_bytes(src), compat.as_bytes(dst), overwrite)
tensorflow.python.framework.errors_impl.AlreadyExistsError: file already exists

```

**Comment by Conchylicultor** [comment]
If you're running in distributed mode, I think you should first generate the dataset in some shared folder (like Google Cloud Storage or similar), then use `data_dir='gs://my/dataset'` to access the dataset.
You also need to be careful that each worker should read a different portion of the dataset, or you will read the same examples multiple times across workers. You can try the split API for this: https://www.tensorflow.org/datasets/splits.

**Comment by Flamefire** [comment]
> I think you should first generate the dataset in some shared folder 

I did that now as I figured the same. It still is strange as this is something done in the background without explicit user interaction. Previously you had a separate command to create the tf_record files, now it is hidden behind the `load` method and that doesn't work with a distributed strategy

> You also need to be careful that each worker should read a different portion of the dataset, or you will read the same examples multiple times across workers.

Shouldn't this be done by the strategy? As I understood it this is (one of its) purposes. There is something like `distribute_dataset` or so called behind the scenes.

**Comment by Conchylicultor** [comment]
You can disable generation by setting `tfds.load(..., download=False)`. Having `download=True` by default was a deliberate choice, even if we understand not everyone might like this default.
Or alternatively, `builder = tfds.builder('my_dataset')` followed by `builder.as_dataset()` won't download the data.

For the multi-worker distribution strategy, `experimental_distribute_dataset` tries to automatically shard the dataset, but I haven't tried if it works well in practice. Note that if you're using `shuffle_files=True` (as you should for training), and `ds = ds.shuffle()`, you need to make sure that the dataset is deterministically generated, by passing seed to `ds.shuffle` and `tfds.ReadConfig()` to `tfds.load`, otherwise, the automatic sharding won't works.
Alternatively, you could have a look at https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy?version=stable#experimental_distribute_datasets_from_function
My suggestion would be to read the doc carefully and test with some dummy dataset to make sure this works correctly (no overlap between datasets)

**Comment by Flamefire** [comment]
> Note that if you're using shuffle_files=True

IIRC there was a warning that this is deprecated, hence my setting to False. Has this been changed?

> s = ds.shuffle(), you need to make sure that the dataset is deterministically generated, by passing seed to ds.shuffle

Oh, good point. Couldn't this be done automatically? The doc for the distribute says

> In a multi-worker setting, we will first attempt to distribute the dataset by attempting to detect whether the dataset is being created out of ReaderDatasets (e.g. TFRecordDataset, TextLineDataset, etc.) and if so, attempting to shard the input files.

This seems to say that the distribution happens at the file level, hence no need to keep the shuffle deterministic. Or did I misunderstood that?

**Comment by Conchylicultor** [comment]
I believe the automatic file sharding only works if called as `TFRecordDataset([file0, file1,...])`.
TFDS uses `ds.interleave()` to construct the multiple files, so I don't think the automatic sharding works in this case.
It is the user responsibility to ensure the pipeline is deterministic. See: https://www.tensorflow.org/api_docs/python/tf/data/Dataset?version=stable#shard

In any case, the least error prone would be to explicitly shard the dataset per worker using the [subsplit API](https://www.tensorflow.org/datasets/splits) and `experimental_distribute_datasets_from_function`

**Comment by Flamefire** [comment]
I think this should "just work": https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras#dataset_sharding_and_batch_size

> In multi-worker training, sharding data into multiple parts is needed to ensure convergence and performance. However, note that in above code snippet, the datasets are directly sent to model.fit() without needing to shard; this is because tf.distribute.Strategy API takes care of the dataset sharding automatically in multi-worker trainings.

I'll see if I can come up with some pseudo-dataset to test this. Not sure how to do that though. TF2 API is quite complicated and documentation isn't that mature yet if you want to do something "more advanced"

**Comment by Conchylicultor** [comment]
From: https://www.tensorflow.org/api_docs/python/tf/data/Dataset?version=stable#shard
`Be sure to shard before you use any randomizing operator (such as shuffle).`

And https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy?version=stable#experimental_distribute_dataset:
`we will shard the dataset evenly at the end by appending a .shard operation to the end of the processing pipeline.`

I think the Keras documentation is incomplete and that it would works only if the dataset yield samples in deterministic order (seed is set or shuffle is disabled)

**Comment by Conchylicultor** [comment]
My feeling would be to use `experimental_distribute_datasets_from_function` which is more explicit, so less error prone and should be more performant:

Something like:
```
def dataset_fn(ctx):
  splits_boundaries = [int(i/ctx.num_input_pipelines * 100) for i in range(ctx.num_input_pipelines + 1)]
  splits = [
      'train[{}%:{}%]'.format(minv, maxv) 
      for minv, maxv in zip(splits_boundaries[:-1], splits_boundaries[1:])
  ]
  ds = tfds.load(..., split=splits[ctx.input_pipeline_id])
  ds = ds.shuffle()
  return ds

ds = strategy.experimental_distribute_datasets_from_function(dataset_fn)
```

**Comment by Flamefire** [comment]
I did a very simple test: Create a dummy dataset which has numbered samples. Then use the standard TF workflow to distribute this to 2 workers and print out the samples seen by each worker. It does work indeed without fiddling with the distribution and all workers (combined) do see each "image" exactly once. So there is no duplication, i.e. each worker does work on a subset of the whole dataset.

See [dummy-distribute.tar.gz](https://github.com/tensorflow/datasets/files/4127513/dummy-distribute.tar.gz)

**Comment by Conchylicultor** [comment]
Thank you for the update. I think it worked because `shuffle_files=False` by default in TFDS. I you were trying with `tfds.load(..., shuffle_files=True)`, my feeling is that two workers may see the same image, unless the user specify the seed explicitly `tfds.ReadConfig(seed=...)`.

But as long as you understand what is happening with determinism and shuffling, you're good to use the auto-sharding.
Is there another issue or can this issue be closed ?

**Comment by Flamefire** [comment]
Yes I think this issue can be closed as "won't fix" unless you intend integration with the TF strategy and make it prepare the dataset only once

**Comment by Conchylicultor** [comment]
Thank you for the confirmation.
I don't think it would be feasible to make tfds distribution aware to generate the data only once. For instance what if each worker is working on different version of the dataset, workers should now what other workers are building to decide if they need to build the dataset or not.
The best thing we could do may be to raise a more explicit error message to forbid generation in a multi-worker setting, but even this is not sure this is desirable.
